cargo-marker/src/config.rs

```diff
@@ -0,0 +1,154 @@
+
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    lints: HashMap<String, LintDependency>,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum LintDependency {
+    /// Version string like: `lint = "0.0.1"`
+    Simple(String),
+    /// Full dependency entry, like:
+    /// lint = { path = ".", version = "1.0.0" }
+    Full(LintDependencyEntry),
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct LintDependencyEntry {
+    path: Option<String>,
+    package: Option<String>,
+    // TODO: Everything below is not yet supported
+    // Registry fetching:
+    version: Option<String>,
+    registry: Option<String>,
+    // Git source fetching:
+    git: Option<String>,
+    rev: Option<String>,
+    branch: Option<String>,
+    // Features:
+    #[serde(rename = "default-features")]
+    default_features: Option<bool>,
+    features: Option<Vec<String>>,
+    optional: Option<bool>,
+    // TODO: do we want lint configuration here too?
+    // configuration: Option<Value>
+}
+
+#[derive(Debug)]
+pub enum ConfigFetchError {
+    /// `Cargo.toml` wasn't found
+    FileNotFound,
+    /// Read failed
+    IoError(io::Error),
+    /// Couldn't parse `Cargo.toml`
+    ParseError(toml_edit::de::Error),
+    /// `workspace.metadata.marker` has invalid structure
+    InvalidStructure,
+    /// `workspace.metadata.marker` doesn't exist
+    NotFound,
+}
+
+impl ConfigFetchError {
+    pub fn emit_and_convert(self) -> ExitStatus {
+        match self {
+            ConfigFetchError::FileNotFound => eprintln!("`Cargo.toml` wasn't found"),
+            ConfigFetchError::IoError(err) => eprintln!("IO error reading config: {err:?}"),
+            ConfigFetchError::ParseError(err) => eprintln!("Can't parse config: {err:?}"),
+            ConfigFetchError::NotFound => eprintln!("Marker config wasn't found"),
+            ConfigFetchError::InvalidStructure => {
+                eprintln!("`workspace.metadata.marker` has invalid structure");
+                return ExitStatus::WrongStructure;
+            },
+        };
+        ExitStatus::BadConfiguration
+    }
+}
+
+macro_rules! unsupported_fields {
+    ($name:expr, $dep:expr => [$($i:ident),+]) => {
+        $(
+            if let Some(ref $i) = $dep.$i {
+                eprintln!(concat!("warning: {} ({:?}): marker doesn't yet support `", stringify!($i), "` field"), $name, $i);
+            }
+        )+
+    }
+}
+
+impl Config {
+    pub fn get_marker_config() -> Result<Config, ConfigFetchError> {
+        let Ok(mut config_file) = File::open(CARGO_TOML) else {
+            return Err(ConfigFetchError::FileNotFound);
+        };
+
+        let mut config_str = String::new();
+        if let Err(e) = config_file.read_to_string(&mut config_str) {
+            return Err(ConfigFetchError::IoError(e));
+        }
+
+        let cargo_config = match from_str::<Value>(&config_str) {
```

Comment:
Just a general comment to think about for the future:

We could split this function, to only the file loading part and the thing that parses the text. The advantage would be testability. So far, there are no real tests for `cargo-marker` and for development it's good enough if it just works. So, you don't have to add them as part of this PR. It's just something we should think about later down the line :)

Reply:
Done